gutenberg/blocks/ts/src/components/drag-and-drop-select-list.tsx

```diff
@@ -0,0 +1,187 @@
+import React from 'react';
+import { wp } from 'wp';
+import MenuIcon from '../menu-icon';
+import "./drag-and-drop-select-list.scss";
+
+declare var wp: wp;
+const { __ } = wp.i18n;
+
+interface State {
+  items: string[];
+  selected: string[];
+}
+
+interface Props {
+  value: string[];
+  forcedItems: string[];
+  optionalItems: string[];
+  onChange: (value: string[]) => void;
+}
+
+class DragAndDropSelectList extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      items: [],
+      selected: []
+    };
+  }
+
+  /**
+   * Component did mount life-cycle event
+   */
+  public componentDidMount = async () => {
+    const selected: any[] = [];
+    const items: string[] = [];
+
+    for (let i = 0; i < this.props.value.length; i++) {
+      selected.push(this.props.value[i]);
+    }
+
+    this.props.forcedItems.forEach(item => {
+      if (!selected.includes(item)) {
+        selected.push(item);
+      }
+    });
+
+    this.props.optionalItems.forEach(item => {
+      if (!selected.includes(item)) {
+        items.push(item);
+      }
+    });
+
+    this.setState({
+      items,
+      selected
+    });
+  };
+
+  draggedItem: any;
+  draggedIdx: any;
+
+  onDragStart = (e: any, index: number) => {
+    this.draggedItem = this.state.selected[index];
+    e.dataTransfer.effectAllowed = 'move';
+    e.dataTransfer.setData('text/html', e.target.parentNode);
+    e.dataTransfer.setDragImage(e.target.parentNode, 20, 20);
+  };
+
+  onDragOver = (index: number) => {
+    const draggedOverItem = this.state.selected[index];
+
+    if (this.draggedItem === draggedOverItem) {
+      return;
+    }
+    let selected = this.state.selected.filter(item => item !== this.draggedItem);
+    selected.splice(index, 0, this.draggedItem);
```

Comment:
What is this "this.draggedItem"

Reply:
It is the item that is currently being dragged on the list, uses "draggedItem: any;" declaration to work